src/ast/passes/semantic_analyser.cpp

```diff
@@ -1225,6 +1238,59 @@ void SemanticAnalyser::visit(Call &call)
             << step->value << " for range " << (max->value - min->value) << ")";
       }
     }
+  } else if (call.func == "tseries") {
+    const static std::unordered_set<std::string> ALLOWED_AGG_FUNCS = {
+      "avg",
+      "sum",
+      "max",
+      "min",
+    };
+    if (is_final_pass()) {
+      Expression &interval_ns_arg = call.vargs.at(3);
+      Expression &num_intervals_arg = call.vargs.at(4);
+
+      auto *interval_ns = interval_ns_arg.as<Integer>();
+      auto *num_intervals = num_intervals_arg.as<Integer>();
+
+      if (!interval_ns) {
+        call.addError()
+            << call.func
+            << ": invalid interval_ns value (must be non-negative literal)";
+        return;
+      }
+      if (!num_intervals) {
+        call.addError()
+            << call.func
+            << ": invalid num_intervals value (must be non-negative literal)";
+        return;
+      }
+
+      if (interval_ns->value <= 0) {
+        call.addError() << "tseries() interval_ns must be >= 1 ("
+                        << interval_ns->value << " provided)";
+        return;
+      }
+
+      if (num_intervals->value <= 0) {
+        call.addError() << "tseries() num_intervals must be >= 1 ("
+                        << num_intervals->value << " provided)";
+        return;
+      } else if (num_intervals->value > 1000000) {
+        call.addError() << "tseries() num_intervals must be < 1000000 ("
+                        << num_intervals->value << " provided)";
+      }
+
+      if (call.vargs.size() == 6) {
+        auto &aggregator = *call.vargs.at(5).as<String>();
+        if (!ALLOWED_AGG_FUNCS.contains(aggregator.value)) {
+          call.addError() << "tseries() expects one of the following "
+                             "aggregation functions: avg, min, max, sum"
```

Comment:
nit: you could generate these err msg values from your set above

Reply:
Done